Switch to PyInstaller to fix build issues

linuxdaemon opened the pull request:
PyInstaller plays nicer with linux, and seems to work fine on mac and windows too, so I switched the builds over to using it for all platforms. There were a few issues at first (macos permissions issues, windows sets sys.stdout to None, etc) but got those all fixed up.

This does introduce one breaking change, it moves the default output directories and the config location to the platform default directories:
- output is `%USERPROFILE%/Videos/videos_with_ad` by default on Windows, `~/Movies/videos_with_ad` on Mac and linux
- Config data goes in `%USERPROFILE%\\AppData\\Local\\describealign` (`~/.config/describealign` on Linux and `~/Library/Application Suppoer/describealign` on Mac) so the config is kept between different downloads of the software

These directory changes were made to keep it consistent across platforms and to fix some more permissions errors on all systems (writing the config file in the `describealign` install folder tends to fail)

This also includes some slight bumps in library versions, as the version of numpy we were running was producing binary incompatibility errors on Windows 10, but upgrading to a new minor version fixed that

linuxdaemon committed:
Test building with pyinstaller

linuxdaemon committed:
Move ls to right place

linuxdaemon committed:
Add linux build dependencies

linuxdaemon committed:
Install linux packages only on linux

linuxdaemon committed:
Remove windows line continuation

linuxdaemon committed:
Handle build artifacts from pyinstaller

linuxdaemon committed:
Fix output dir for macos .app

linuxdaemon committed:
Use platformdirs to generate default paths

This ensures the paths are consistent with other apps, and that, for example, the config file is not stored in the .app bundle (which has the wrong file permissions)

linuxdaemon committed:
Fix windows artifact path

linuxdaemon committed:
Try macos homedir install to fix permissions issue

linuxdaemon committed:
Disable homedir install, add .spec artifcts, install ffmpeg on mac

linuxdaemon committed:
Add spec file for build

linuxdaemon committed:
Set exec bit on script

linuxdaemon committed:
Allow running describe-align with only --install-ffmpeg

linuxdaemon committed:
Handle no args early in execution

linuxdaemon committed:
Fix positional arguments

linuxdaemon committed:
Remove second static_ffmpeg import

linuxdaemon committed:
Fix install location for macos package

linuxdaemon committed:
Fix ffmpeg exec bits

linuxdaemon committed:
Fix ffprobe permissions too

linuxdaemon committed:
Fix type errors concatting with non-strings

linuxdaemon committed:
Bump lib versions

linuxdaemon committed:
Redirect stdout to fix errors on windows

linuxdaemon committed:
Remove duplicate logging, fix newline handling in status console

linuxdaemon committed:
Fix windows path name

linuxdaemon committed:
Zip windows dist dir

linuxdaemon committed:
Fix zip on windows

linuxdaemon committed:
Add run condition for zip step

linuxdaemon committed:
Remove nuitka config

julbean commented on .github/workflows/build-binary.yaml:
```diff
@@ -24,32 +24,38 @@ jobs:
-          pip install imageio
+          pip install -Ur requirements.txt
+          pip install -U pillow
+          pip install -U pyinstaller
```
should these be specific versions rather than the latest?

linuxdaemon replied:
It may make sense to define these with version constraints yeah, I'll move them to their own build-requirements.txt file and set ranges

linuxdaemon replied:
Done

julbean commented on describealign.py:
```diff
@@ -1125,7 +1126,7 @@ def combine_gui(video_files, audio_files, config_path):
 
 def main_gui():
-  config_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'config.ini')
+  config_path = platformdirs.user_config_path(appname='describealign', ensure_exists=True) / 'config.ini'
```
I can imagine the old method here having problems, but I don't think I ever saw any issues even with the Mac binaries.  Where were you seeing a problem with writing the config?  If we are gonna change this, I'd rather not break existing installs by resetting their configs when they upgrade with pip.  But that could be solved by checking the old config location and moving it over if there isn't a new one already.

linuxdaemon replied:
On mac, it would attempt to write the config in the installed app dir /Applications/describealign.app which is owned by root. We can definitely add checks for the old location yeah

linuxdaemon replied:
Added migration function to move the old config

julbean commented on describealign.py:
```diff
@@ -995,7 +996,7 @@ def read_config_file(config_path):
   return settings
 
-def settings_gui(config_path):
+def settings_gui(config_path: Path):
```
Why the type hint here specifically?

linuxdaemon replied:
This was to make it clear what type is being passed, as it used to be a string but with the new platformdirs call, its a Path object and some functions don't work the same (ie string concat)

julbean commented on describealign.py:
```diff
@@ -99,15 +101,14 @@
 if IS_RUNNING_WINDOWS:
   import PySimpleGUIWx as sg
-  default_output_dir = 'videos_with_ad'
-  default_alignment_dir = 'alignment_plots'
```
Are these broken?  I can't say I was ever happy with the Mac implementation I slapped together here (originally to fix issues with the binary), but I think we should have a good reason for changing the default output directories.

linuxdaemon replied:
Mainly this was a consistency change to keep the dirs the same even if you reinstall/redownload the build (and possibly run it from a different dir). I can definitely change these back if you'd prefer though

julbean replied:
With the new changes you made to the config, the directories should be maintained across installs so long as the user manually set them.  Otherwise, I think relative path behavior is more intuitive as a default than sending stuff to the Videos folder.  I don't think it's a big difference, though, so I'd be down to change it if you feel strongly, but when combined with the inertia of making a breaking change, barring something else I'm missing here, I don't think it's worth it.

linuxdaemon replied:
Good point, I'll revert this for now, the $HOME based paths on mac+linux do solve the issues

julbean commented on describealign.py:
```diff
@@ -1233,8 +1235,15 @@ def error(self, message):
   args = parser.parse_args()
   
+  if args.install_ffmpeg:
+    static_ffmpeg.run.get_or_fetch_platform_executables_else_raise()
```
I don't remember exactly what it was, but there was some issues with the Filelock dependency this version of this command.  That's why it's the non-locking version elsewhere

linuxdaemon replied:
This was already changed in a later commit to use the existing get_ffmpeg and get_ffprobe functions in describealign

julbean commented on Package/scripts/postinstall:
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
+exec "$DSTROOT/describealign.app/Contents/MacOS/describealign" --install-ffmpeg
```
This seems pretty unnecessary to me.  What's the benefit here?  Won't this be installed on first run?

linuxdaemon replied:
With the mac installer, the files end up being owned by root as they are installed in /Applications/describealign.app, so if it tries to install on first run it will fail with permissions errors. `postinstall` gets run in the context of the installer which ensures it has the proper permissions to write to those paths

julbean commented on describealign.py:
```diff
@@ -1187,20 +1187,11 @@ def main_gui():
 # > describealign video.mp4 audio_desc.mp3
 def command_line_interface():
-  # override command line argument parser's error handler to make it pause before exiting
-  # this allows users to see the error message when accidentally not running from command line
```
Why are we removing this?  I like this feature.  Does the default=None on the inputs somehow fix this?

julbean replied:
Okay, I think I see what you're doing.  The next commit together with this is making this whole thing cleaner

linuxdaemon replied:
It's not being removed, just changed so that if any args are specified it won't open the gui. Currently if you try `describealign --stretch_audio` for instance, instead of printing the command usage it opens the GUI. With the new check `if len(sys.argv) < 2:` it will only open the GUI if the app is run with no args

julbean commented on describealign.py:
```diff
@@ -1187,6 +1187,11 @@ def main_gui():
 # > describealign video.mp4 audio_desc.mp3
 def command_line_interface():
+  if len(sys.argv) < 2:
+    # No args, run gui
```
Should probably add the print back in here

linuxdaemon replied:
Will do

julbean commented on describealign.py:
```diff
@@ -1195,8 +1195,8 @@ def command_line_interface():
                           usage="describealign video_file.mp4 audio_file.mp3")
-  parser.add_argument("video", help='A video file or directory containing video files.', default=None)
-  parser.add_argument("audio", help='An audio file or directory containing audio files.', default=None)
+  parser.add_argument("video", help='A video file or directory containing video files.', nargs='?')
```
Does nargs='?' work without a default value?  The docs don't seem to discuss what behavior that has: https://docs.python.org/3/library/argparse.html#nargs

linuxdaemon replied:
```python
>>> import argparse
>>> parser = argparse.ArgumentParser()
>>> parser.add_argument('foo', nargs='?')
_StoreAction(option_strings=[], dest='foo', nargs='?', const=None, default=None, type=None, choices=None, required=False, help=None, metavar=None)
>>> parser.parse_args([])
Namespace(foo=None)
>>> parser.parse_args(['bar'])
Namespace(foo='bar')
```

The default for any argument is `None` unless otherwise specified with `default=`, and setting `default=` on a positional argument is an error.

julbean replied:
Setting the default value as None doesn't give an error for me:
```
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('blah', nargs='?', help='Blah.', default=None)
parser.add_argument('blah2', nargs='?', help='Blah2.', default=None)

print(parser.parse_args(['in', 'in2']))
print(parser.parse_args(['in']))
print(parser.parse_args([]))
```
prints:
```
Namespace(blah='in', blah2='in2')
Namespace(blah='in', blah2=None)
Namespace(blah=None, blah2=None)
```
I'd rather be explicit about the default value, but not a big deal if you think it's cleaner without it

linuxdaemon replied:
Ah yep I misremembered my testing, setting `optional` fails, not `default`. Yep, 100% agree with being more explicit, thanks for catching this!

julbean commented on describealign.py:
```diff
@@ -83,14 +83,12 @@
 import os
 import glob
 import itertools
-import datetime
```
Wonder what I had been using this for haha

linuxdaemon replied:
idk, it was unused and pyinstaller was including it in the package so I dropped it lol

julbean commented on describealign.py:
```diff
@@ -1235,7 +1236,8 @@ def command_line_interface():
   if args.install_ffmpeg:
-    static_ffmpeg.run.get_or_fetch_platform_executables_else_raise()
+    # Make sure the file is world executable
+    os.chmod(get_ffmpeg(), 0o755)
```
Nice

linuxdaemon replied:
This ensures the file has the correct perms, as for some reason on macos it ends up with `rwxr-x--x root wheel`, which does not allow execution as a non-root user despite having `x`.

julbean commented on describealign.py:
```diff
@@ -1076,10 +1077,21 @@ def settings_gui(config_path: Path):
   try:
-    combine(*args, **kwargs)
-  except:
+    with redirect_stdout(writer), redirect_stderr(writer):
```
Interesting method.  Is this functioning as an alternative to my exception printing hack?  Then why do we still have that?  I must be missing something here

julbean commented on describealign.py:
```diff
@@ -1088,12 +1088,11 @@ def write(self, s: str) -> int:
-  except Exception:
-    print_queue.put(traceback.format_exc())
-    # raise
+  with redirect_stdout(writer), redirect_stderr(writer):
```
Ah that makes more sense, haha

julbean commented on describealign.py:
```diff
@@ -1117,7 +1116,7 @@ def combine_gui(video_files, audio_files, config_path):
       if IS_RUNNING_WINDOWS:
         cursor_position = output_textbox.WxTextCtrl.GetInsertionPoint()
-      output_textbox.update(f'\n {print_queue.get()}', append=True)
+      output_textbox.update(print_queue.get(), append=True)
```
Is this cleaner version simply equivalent or does it also fix a bug?

linuxdaemon replied:
As part of the above change to just capturing `print`, we already have the `\n`, so this adds a second one. And when static_ffmpeg is printing `.` to denote progress, it calls sys.stdout.write which means a new call to queue.put(), so it was spamming the GUI console with every `.` on a new line.

julbean commented on describealign.py:
```diff
@@ -1105,7 +1104,7 @@ def combine_gui(video_files, audio_files, config_path):
   
   settings = read_config_file(config_path)
-  settings.update({'display_func':print_queue.put, 'yes':True})
+  settings.update({'yes':True})
```
Oh looks like I never used this

linuxdaemon replied:
Nah display_func was being passed to `combine` to allow logging to the console with `display`, but with the use of redirect_stdout/redirect_stderr, all calls to print() are being redirected to the GUI console. We can transition away from using `display`.

I switched to redirect_stdout because by default on Windows (with PyInstaller) sys.stdout is None, and static_ffmpeg calls sys.stdout.write directly in several places, which was initially crashing.

linuxdaemon committed:
Set version constraints on build dependencies

linuxdaemon committed:
Migrate config if existing config found

julbean commented:
Apparently a bunch of my comments got randomly marked as "Pending" for some reason.  Thanks github

julbean commented on describealign.py:
```diff
@@ -1109,8 +1109,27 @@ def combine_gui(video_files, audio_files, config_path):
+  
+  old_data = old_path.read_text(encoding='utf-8')
+  new_path.write_text(old_data, encoding='utf-8')
+  print(f"Configuration migrated to {new_path}")
```
Should we also delete the old config?

linuxdaemon replied:
My concern is that the old config might not be writable, only readable to the current user (should be rare), but since I added the error check, I can add a delete yeah, should be fine.

linuxdaemon committed:
Revert back to original default output dirs

linuxdaemon committed:
Add back the print when launching GUI

linuxdaemon committed:
Delete old config after migration

linuxdaemon committed:
Use the correct form of format_exception_only

linuxdaemon committed:
Switch QueueWriter to inherit TextIOWrapper

This ensures it implements all of the methods expected of sys.stdout, it shouldn't really make a difference but this is more type-safe

linuxdaemon committed:
Set default explicitly

julbean commented:
Mac artifact got through 70% of the install then gave me the generic "Installation failed, encountered an error" message.  Windows one seems to be double-zipped, but appears to run fine.  config.ini file also ends up in a nested describealign directory in AppData/Local.  Running it in console still doesn't print anything when launching the GUI, but does launch it.  That's an improvement over a bug in the live binary, which just crashes without printing anything.  Running it from command line with input files still fails (like the live binary), but produces an "Unhandled exception" popup now:
```
  File "describealign.py", line 1265, in <module>
  File "describealign.py", line 1254, in command_line_interface
  File "describealign.py", line 855, in combine
RuntimeError: input(): lost sys.stdin
```
Trying the GUI on files also spits out a Deprecation Warning:
```
describealign.py:500: MatplotlibDeprecationWarning: The legendHandles attribute was deprecated in Matplotlib 3.7 and will be removed two minor releases later. Use legend_handles instead.
```
Also pops up gross command prompt windows while ffmpeg runs.
The code cleanup removing manual line breaks seems to have messed up manual messages like this one:
```
Combining media files:Directory not found, creating it:
```
So we'll need to add a line break to the end of that one and any others like it.

Linux zip contains a .tar.gz.  Running the binary does open the GUI and it runs properly, installing ffmpeg and everything, up until it prints "processing files" and just stops there.  It doesn't use any CPU and clicking "Close" from there freezes the GUI.  After force quitting and retrying, it worked perfectly the second time, also worked fine even after deleting the folder and starting from scratch.
